lms/static/scripts/frontend_apps/components/ErrorDisplay.js

```diff
@@ -30,82 +30,114 @@ function toSentence(str) {
- */
-
-/**
- * Displays details of an error, such as a failed API call and provide the user
- * with information on how to get help with it.
+ * JSON-stringify `error.details` if it is extant and an object
  *
- * @param {ErrorDisplayProps} props
+ * @param {ErrorLike} error
  */
-export default function ErrorDisplay({ message, error }) {
-  /** @type {string|undefined} */
-  let details = '';
-
-  if (typeof error.details === 'object' && error.details !== null) {
+function formatErrorDetails(error) {
+  /** @type {string|object} */
+  let details = error.details ?? '';
+  if (error?.details && typeof error.details === 'object') {
     try {
       details = JSON.stringify(error.details, null, 2 /* indent */);
     } catch (e) {
       // ignore
     }
-  } else {
-    details = error.details;
   }
+  return details;
+}
 
-  let supportEmailBody = '';
-  if (error.message) {
-    supportEmailBody += `\n\nError message: ${error.message}`;
-  }
-  if (details) {
-    supportEmailBody += `\n\nTechnical details:\n\n${details}`;
+/**
+ * @typedef ErrorDetailsProps
+ * @prop {ErrorLike} error
+ *
+ * Render pre-formatted JSON details of an error
+ *
+ * @param {ErrorDetailsProps} props
+ */
+function ErrorDetails({ error }) {
+  /** @param {Event} event */
+  const onDetailsToggle = event => {
+    const detailsEl = /** @type {HTMLDetailsElement} */ (event.target);
+    if (!detailsEl.open) {
+      return;
+    }
+    detailsEl.scrollIntoView({ behavior: 'smooth' });
+  };
+
+  const details = formatErrorDetails(error);
+  if (!details) {
+    return null;
   }
 
+  return (
+    <details className="ErrorDisplay__details" onToggle={onDetailsToggle}>
+      <summary className="ErrorDisplay__details-summary">Error Details</summary>
+      <pre className="ErrorDisplay__details-content">{details}</pre>
+    </details>
+  );
+}
+
+/**
+ * @typedef ErrorDisplayProps
+ * @prop {string|null} [description] -
+ *   A short message explaining the error and its human-facing relevance.
+ *   This is typically a general message like
+ *   "There was a problem fetching this assignment". This description
+ *   always comes from (this) client app.
+ *
+ *   The presence of a `description` indicates that this `ErrorDisplay` is
+ *   responsible for explaining the error to the user in some fashion. Along
+ *   with this `description`, anything available in `error.message` is also
+ *   shown.
+ *
+ *   When `description` is absent, it indicates that the main explaining of
+ *   the error's relevance to the user is handled elsewhere, and no additional
+ *   error messaging is necessary. This is the case, for example, with some
+ *   of the well-explained error states in `OAuth2RedirectErrorApp` or
+ *   `LaunchErrorDialog`: these don't need additional error messaging that is,
+ *   for the most part, redundant or useless.
+ *
+ *   Available `error.details` and support/email instructions are always
+ *   shown regardless of the presence of this prop.
```

Comment:
The `After` screenshot of `OAuth2RedirectErrorApp` in the description of the PR doesn't have the `send us an email` option:

![image](https://user-images.githubusercontent.com/439947/135139959-5f9dc8f6-56e4-4e46-8825-f66e45b3e95d.png)

Reply:
Sorry, this screenshot is misleading. Because this modal gets so tall here when the details are expanded, it becomes scrollable. The instructions _are_ present, but they're scrolled up/not visible here.

Subsequent work will involve replacing these modals with shared variants, which will allow for tuning the scrollable approach.